File: website/tutorials/random/ImprovedRandom.cpp
#include <stdlib.h>
#include <time.h>

#include "hlt.hpp"
#include "networking.hpp"

int main() {
    srand(time(NULL));

    std::cout.sync_with_stdio(0);

    unsigned char myID;
    hlt::GameMap presentMap;
    getInit(myID, presentMap);
    sendInit("C++Bot");

    std::set<hlt::Move> moves;
    while(true) {
        moves.clear();

        getFrame(presentMap);

        for(unsigned short a = 0; a < presentMap.height; a++) {
            for(unsigned short b = 0; b < presentMap.width; b++) {
                if (presentMap.getSite({ b, a }).owner == myID) {

                    bool movedPiece = false;
                    
                    for(int d : CARDINALS) {
                        if(presentMap.getSite({ b, a }, d).owner != myID && presentMap.getSite({ b, a }, d).strength < presentMap.getSite({ b, a }).strength) {
                            moves.insert({ { b, a }, d });
                            movedPiece = true;
                            break;
                        }
                    }

                    if(!movedPiece && presentMap.getSite({ b, a }).strength < presentMap.getSite({ b, a }).production * 5) {
                        moves.insert({ { b, a }, STILL });
                        movedPiece = true;
                    }
                    
                    if(!movedPiece) {
                        moves.insert({ { b, a }, (bool)(rand() % 2) ? NORTH : WEST });
                        movedPiece = true;
                    }
                }
            }
        }

        sendFrame(moves);
    }

    return 0;
}

File: website/tutorials/random/ImprovedRandom.py
from hlt import *
from networking import *

myID, gameMap = getInit()
sendInit("PythonBot")

while True:
    moves = []
    gameMap = getFrame()
    for y in range(gameMap.height):
        for x in range(gameMap.width):
            if gameMap.getSite(Location(x, y)).owner == myID:

                movedPiece = False

                for d in CARDINALS:
                    if gameMap.getSite(Location(x, y), d).owner != myID and gameMap.getSite(Location(x, y), d).strength < presentMap.getSite(Location(x, y)).strength:
                        moves.append(Move(Location(x, y), d))
                        movedPiece = True
                        break

                if not movedPiece and gameMap.getSite(Location(x, y)).strength < gameMap.getSite(Location(x, y)).production * 5:
                    moves.append(Move(Location(x, y), STILL))
                    movedPiece = True

                if not movedPiece:
                    moves.append(Move(Location(x, y), NORTH if bool(int(random.random() * 2)) else WEST))
                    movedPiece = True

    sendFrame(moves)